File: autocontext/src/autocontext/research/consultation.py
"""Research consultation — goal decomposition and brief assembly (AC-499).

Domain service: ResearchConsultant decomposes a goal into targeted queries,
executes them through a ResearchEnabledSession, filters weak signals, deduplicates
citations, and packages everything into a ResearchBrief value object.

ResearchBrief is a frozen value object suitable for downstream prompt injection.
"""

from __future__ import annotations

import logging
from collections.abc import Sequence

from pydantic import BaseModel, Field

from autocontext.research.runtime import ResearchEnabledSession
from autocontext.research.types import Citation, ResearchQuery, ResearchResult, Urgency

logger = logging.getLogger(__name__)


class ResearchBrief(BaseModel):
    """Immutable snapshot of research findings for a goal.

    Produced by ResearchConsultant, consumed by prompt wiring (AC-501).
    """

    goal: str
    findings: list[ResearchResult] = Field(default_factory=list)
    unique_citations: list[Citation] = Field(default_factory=list)

    model_config = {"frozen": True}

    @property
    def avg_confidence(self) -> float:
        if not self.findings:
            return 0.0
        return sum(f.confidence for f in self.findings) / len(self.findings)

    @classmethod
    def from_results(
        cls,
        goal: str,
        results: Sequence[ResearchResult],
        min_confidence: float = 0.0,
    ) -> ResearchBrief:
        filtered = [r for r in results if r.confidence >= min_confidence]
        citations = _dedupe_citations(filtered)
        return cls(goal=goal, findings=list(filtered), unique_citations=citations)

    @classmethod
    def empty(cls, goal: str) -> ResearchBrief:
        return cls(goal=goal)

    def to_markdown(self) -> str:
        if not self.findings:
            return f"## Research Brief: {self.goal}\n\nNo findings available.\n"

        parts = [f"## Research Brief: {self.goal}\n"]
        for f in self.findings:
            parts.append(f"### {f.query_topic} (confidence: {f.confidence:.0%})\n")
            parts.append(f"{f.summary}\n")
            for c in f.citations:
                label = f"[{c.source}]({c.url})" if c.url else c.source
                parts.append(f"- {label}")
                if c.snippet:
                    parts.append(f"  > {c.snippet}")
            parts.append("")
        return "\n".join(parts)


def _dedupe_citations(results: Sequence[ResearchResult]) -> list[Citation]:
    """Collect unique citations across results, keyed by (source, url)."""
    seen: set[tuple[str, str]] = set()
    unique: list[Citation] = []
    for r in results:
        for c in r.citations:
            key = (c.source, c.url)
            if key not in seen:
                seen.add(key)
                unique.append(c)
    return unique


class ResearchConsultant:
    """Domain service: decompose goal → queries → brief.

    Stateless — create one and call .consult() per research need.
    """

    def __init__(
        self,
        urgency: Urgency = Urgency.NORMAL,
        min_confidence: float = 0.0,
    ) -> None:
        self._urgency = urgency
        self._min_confidence = min_confidence

    def consult(
        self,
        session: ResearchEnabledSession,
        topics: Sequence[str],
        context: str = "",
    ) -> ResearchBrief:
        """Execute research queries and return a packaged brief.

        Respects the session's budget — stops when budget is exhausted.
        Filters results below min_confidence.
        """
        if not session.has_research:
            logger.debug("No research adapter attached — returning empty brief")
            return ResearchBrief.empty(session.goal)

        results: list[ResearchResult] = []
        for topic in topics:
            query = ResearchQuery(
                topic=topic,
                context=context,
                urgency=self._urgency,
            )
            result = session.research(query)
            if result is None:
                logger.debug("Budget exhausted after %d queries", len(results))
                break
            results.append(result)

        return ResearchBrief.from_results(
            goal=session.goal,
            results=results,
            min_confidence=self._min_confidence,
        )

File: ts/src/research/consultation.ts
/**
 * Research consultation — goal decomposition and brief assembly (AC-499 TS parity).
 */

import { ResearchEnabledSession } from "./runtime.js";
import { Citation, ResearchQuery, ResearchResult, type Urgency, Urgency as UrgencyEnum } from "./types.js";

function dedupeCitations(results: ResearchResult[]): Citation[] {
  const seen = new Set<string>();
  const unique: Citation[] = [];
  for (const r of results) {
    for (const c of r.citations) {
      const key = `${c.source}||${c.url}`;
      if (!seen.has(key)) { seen.add(key); unique.push(c); }
    }
  }
  return unique;
}

export class ResearchBrief {
  readonly goal: string;
  readonly findings: ResearchResult[];
  readonly uniqueCitations: Citation[];

  constructor(goal: string, findings: ResearchResult[], uniqueCitations: Citation[]) {
    this.goal = goal;
    this.findings = findings;
    this.uniqueCitations = uniqueCitations;
  }

  get avgConfidence(): number {
    if (!this.findings.length) return 0;
    return this.findings.reduce((sum, f) => sum + f.confidence, 0) / this.findings.length;
  }

  static fromResults(goal: string, results: ResearchResult[], minConfidence = 0): ResearchBrief {
    const filtered = results.filter((r) => r.confidence >= minConfidence);
    return new ResearchBrief(goal, filtered, dedupeCitations(filtered));
  }

  static empty(goal: string): ResearchBrief { return new ResearchBrief(goal, [], []); }

  toMarkdown(): string {
    if (!this.findings.length) return `## Research Brief: ${this.goal}\n\nNo findings available.\n`;
    const parts = [`## Research Brief: ${this.goal}\n`];
    for (const f of this.findings) {
      parts.push(`### ${f.queryTopic} (confidence: ${Math.round(f.confidence * 100)}%)\n`);
      parts.push(`${f.summary}\n`);
      for (const c of f.citations) {
        parts.push(c.url ? `- [${c.source}](${c.url})` : `- ${c.source}`);
        if (c.snippet) parts.push(`  > ${c.snippet}`);
      }
      parts.push("");
    }
    return parts.join("\n");
  }

  toJSON(): Record<string, unknown> {
    return { goal: this.goal, findings: this.findings.map((f) => f.toJSON()), uniqueCitations: this.uniqueCitations.map((c) => c.toJSON()) };
  }

  static fromJSON(data: Record<string, unknown>): ResearchBrief {
    const findings = ((data.findings as Record<string, unknown>[]) ?? []).map(ResearchResult.fromJSON);
    const cites = ((data.uniqueCitations as Record<string, unknown>[]) ?? []).map(Citation.fromJSON);
    return new ResearchBrief(data.goal as string, findings, cites);
  }
}

export class ResearchConsultant {
  private _urgency: Urgency;
  private _minConfidence: number;

  constructor(opts?: { urgency?: Urgency; minConfidence?: number }) {
    this._urgency = opts?.urgency ?? UrgencyEnum.NORMAL;
    this._minConfidence = opts?.minConfidence ?? 0;
  }

  consult(session: ResearchEnabledSession, topics: string[], context = ""): ResearchBrief {
    if (!session.hasResearch) return ResearchBrief.empty(session.goal);

    const results: ResearchResult[] = [];
    for (const topic of topics) {
      const result = session.research(new ResearchQuery({ topic, context, urgency: this._urgency }));
      if (!result) break;
      results.push(result);
    }
    return ResearchBrief.fromResults(session.goal, results, this._minConfidence);
  }
}
